appcelerator/titanium_mobile
# refactor(android): ProgressIndicator dialog handling

**JIRA:**
- https://jira.appcelerator.org/browse/TIMOB-27104
- https://jira.appcelerator.org/browse/TIMOB-27308
- https://jira.appcelerator.org/browse/TIMOB-27309

**Summary:**
- [TIMOB-27104](https://jira.appcelerator.org/browse/TIMOB-27104) Fixed bug where `ProgressIndicator` logs "WindowLeaked" exception when hiding dialog and closing window at same time as of 8.0.2.
- [TIMOB-27308](https://jira.appcelerator.org/browse/TIMOB-27308) Fixed bug where determinant `ProgressIndicator` ignores "value" property before shown.
- [TIMOB-27309](https://jira.appcelerator.org/browse/TIMOB-27309) Fixed bug where you cannot re-show `ProgressIndicator` dialog if auto-closed by previous window.
- Removed dialog's `handleMessage()` related code. No longer needed since JS runs on main UI thread.

**Note:**
The status bar `ProgressIndicator` is a dead feature. Google no longer supports it and their Java APIs effectively no-op. I've written up ticket [TIMOB-27312](https://jira.appcelerator.org/browse/TIMOB-27312) to refactor it in the future. In the meantime, there is no point in testing it. _(I've already done some testing on it anyways with this PR and it's no worse.)_

---
**Test 1:**
_(Verifies dialog no longer logs exception when window closes.)_
1. Build and run the code attached to [TIMOB-27104](https://jira.appcelerator.org/browse/TIMOB-27104) on Android.
2. Tap on the "Show Progress Dialog" button.
3. Wait 2 seconds for the progress dialog and its window to close.
4. Verify that a "WindowLeaked" exception was **NOT** logged.

---
**Test 2:**
_(Verifies progress indicator reads "value" property when shown.)_
1. Build and run the code attached to [TIMOB-27308](https://jira.appcelerator.org/browse/TIMOB-27308) on Android.
2. Tap on the "Show Progress Dialog" button.
3. Verify that the dialog shows `50%`.

---
**Test 3:**
_(Verifies progress dialog can be re-shown after auto-closed.)_
1. Build and run the code attached to [TIMOB-27309](https://jira.appcelerator.org/browse/TIMOB-27309) on Android.
2. Tap on the "Show Progress Dialog" button.
3. Wait 2 seconds for the progress dialog and its window to close.
4. Tap on the "Show Progress Dialog" button again.
5. Verify that the progress dialog is re-shown.

---
**Test 4:**
_(Verifies determinant progress indicator handling.)_
1. Build and run with the below code on Android.
2. Tap on the "Show Progress Dialog" button.
3. Verify that a progress dialog is shown.
4. Verify that its progress bar slowly fills from `0%` to `100%` and then closes itself.
5. Tap on the "Show Progress Dialog" button.
6. Tap on the Android "back" button to close the progress dialog.
7. Tap on the "Show Progress Dialog" button.
8. Verify dialog is shown again and starts at `0%`.
9. Verify dialog auto-closes itself when it reaches `100%`.

```javascript
var timerId = null;
var progressIndicator = Ti.UI.Android.createProgressIndicator({
	message: "Progressing...",
	location: Ti.UI.Android.PROGRESS_INDICATOR_DIALOG,
	type: Ti.UI.Android.PROGRESS_INDICATOR_DETERMINANT,
	cancelable: true,
	min: 0,
	max: 100,
});
progressIndicator.addEventListener("cancel", function() {
	if (timerId) {
		clearInterval(timerId);
		timerId = null;
	}
});

var window = Ti.UI.createWindow();
var showButton = Ti.UI.createButton({
	title: "Show Progress Dialog",
	bottom: "30dp",
});
showButton.addEventListener("click", function() {
	progressIndicator.value = 0;
	progressIndicator.show();
	timerId = setInterval(function() {
		var value = progressIndicator.value;
		if (value >= progressIndicator.max) {
			progressIndicator.hide();
			clearInterval(timerId);
			timerId = null;
		} else {
			if (!value) {
				value = progressIndicator.min;
			}
			value += (progressIndicator.max - progressIndicator.min) / 10;
			progressIndicator.value = value;
		}
	}, 250);
});
window.add(showButton);
window.open();
```
